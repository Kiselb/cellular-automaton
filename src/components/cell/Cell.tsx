import React from "react";
import chroma from "chroma-js";

import "./cell.css";

export type CellParams = {
  generation: number;
  row: number;
  col: number;
  colorGamma: string;
  colorEmpty: string;
  colorCell: string;
};

type CellProps = Omit<CellParams, "colorCell"> & {
  onClick: (cell: CellParams) => void;
  probe?: ({}) => void;
};

type CalcCellColor = (colorGamma: string, generation: number) => string;

export const calcCellColor1: CalcCellColor = (
  colorGamma: string,
  generation: number
) => chroma(colorGamma).darken(1 + (generation - 1) * 0.1).hex();

export const calcCellColor: CalcCellColor = (
  colorGamma: string,
  generation: number
) => chroma.scale(['yellow', 'darkgreen'])((generation - 1) * 0.1).hex();

export const Cell = ({
  generation,
  row,
  col,
  onClick,
  colorGamma,
  colorEmpty,
  ...props
}: CellProps) => {
  const colorCell: string = calcCellColor(colorGamma, generation);
  const cellParams: CellParams = {
    generation,
    row,
    col,
    colorGamma,
    colorEmpty,
    colorCell,
  };
  const onClickHandler = (event: React.MouseEvent<HTMLDivElement>) => {
    event.preventDefault();
    onClick(cellParams);
  };
  const onMouseOverHandler = (event: React.MouseEvent<HTMLDivElement>) => {
    event.preventDefault();
    event.buttons === 1 && onClick(cellParams);
    event.buttons === 1 && props.probe && props.probe(cellParams);
  };
  const onDoubleClickHandler = (event: React.MouseEvent<HTMLDivElement>) => {
    event.preventDefault();
    props.probe && props.probe(cellParams);
  };
  return (
    <div
      key={`R${row}:C${col}`}
      data-testid={`R${row}:C${col}`}
      className={["cell"].join(" ")}
      onClick={onClickHandler}
      onDoubleClick={onDoubleClickHandler}
      onMouseOver={onMouseOverHandler}
      style={{ background: generation === 0 ? colorEmpty : colorCell }}
    ></div>
  );
};
