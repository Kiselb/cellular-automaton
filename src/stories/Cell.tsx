import React, { useState } from "react";
import "./cell.css";

export type CellStatus = "dead" | "alive" | "old";

export interface CellParams {
  status: CellStatus;
  row: number;
  col: number;
}

interface CellProps {
  status: CellStatus;
  row: number;
  col: number;
  onClick: (cell: CellParams) => void;
}

export const Cell = ({
  status = "dead",
  row,
  col,
  onClick,
  ...props
}: CellProps) => {
  const onClickHandler = (event: React.MouseEvent<HTMLDivElement>) => {
    event.preventDefault();
    onClick({
      status,
      row,
      col,
    });
  };
  const onMouseOverHandler = (event: React.MouseEvent<HTMLDivElement>) => {
    event.preventDefault();
    if (event.buttons === 1) {
      onClick({
        status,
        row,
        col,
      });
    }
  };
  return (
    <div
      id={"R" + row + "C" + col}
      className={["cell", `cell--${status}`].join(" ")}
      onClick={onClickHandler}
      onMouseOver={onMouseOverHandler}
      {...props}
    ></div>
  );
};
