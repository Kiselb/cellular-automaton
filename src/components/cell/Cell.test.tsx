import React from "react";
import { render, fireEvent, screen } from "@testing-library/react";

import "@testing-library/jest-dom";
import { Cell, calcCellColor } from "./Cell";

it("Cell renders correctly", () => {
  const { asFragment } = render(
    <Cell
      row={1}
      col={1}
      generation={1}
      onClick={() => console.log("OK")}
      colorEmpty="#ececec"
      colorGamma="lightgreen"
    />
  );
  screen.debug();
  expect(asFragment()).toMatchSnapshot();
});
it("Cell has style", () => {
  render(
    <Cell
      row={1}
      col={1}
      generation={1}
      onClick={() => console.log("OK")}
      colorEmpty="#ececec"
      colorGamma="lightgreen"
    />
  );
  const div = screen.getByTestId("R1:C1");
  expect(div).toHaveClass("cell");
});
it("Cell has style background color generation 0", () => {
  render(
    <Cell
      row={1}
      col={1}
      generation={0}
      onClick={() => console.log("OK")}
      colorEmpty="#ececec"
      colorGamma="lightgreen"
    />
  );
  const div = screen.getByTestId("R1:C1");
  expect(div).toHaveStyle({ background: "#ececec" });
});
it("Cell has style background color generation 1", () => {
  render(
    <Cell
      row={1}
      col={1}
      generation={1}
      onClick={() => console.log("OK")}
      colorEmpty="#ececec"
      colorGamma="lightgreen"
    />
  );
  const div = screen.getByTestId("R1:C1");
  expect(div).toHaveStyle({ background: calcCellColor("lightgreen", 1) });
});
it("Cell has style background color generation 10", () => {
  render(
    <Cell
      row={1}
      col={1}
      generation={10}
      onClick={() => console.log("OK")}
      colorEmpty="#ececec"
      colorGamma="lightgreen"
    />
  );
  const div = screen.getByTestId("R1:C1");
  expect(div).toHaveStyle({ background: calcCellColor("lightgreen", 10) });
});
it("Cell has style background color generation 20", () => {
  render(
    <Cell
      row={1}
      col={1}
      generation={20}
      onClick={() => console.log("OK")}
      colorEmpty="#ececec"
      colorGamma="lightgreen"
    />
  );
  const div = screen.getByTestId("R1:C1");
  expect(div).toHaveStyle({ background: calcCellColor("lightgreen", 20) });
});
it("Cell mouse click event", () => {
  const onClick = jest.fn();
  render(
    <Cell
      row={1}
      col={1}
      generation={1}
      onClick={onClick}
      colorEmpty="#ececec"
      colorGamma="lightgreen"
    />
  );
  const div = screen.getByTestId("R1:C1");
  fireEvent.click(div);
  expect(onClick).toBeCalled();
});
it("Cell mouse double click event", () => {
  const probe = jest.fn();
  render(
    <Cell
      row={1}
      col={1}
      generation={1}
      onClick={() => console.log("OK")}
      colorEmpty="#ececec"
      colorGamma="lightgreen"
      probe={probe}
    />
  );
  const div = screen.getByTestId("R1:C1") as HTMLDivElement;
  fireEvent.doubleClick(div);
  expect(probe).toBeCalled();
});
it("Cell mouse over event with pressed button", () => {
  const probe = jest.fn();
  render(
    <Cell
      row={1}
      col={1}
      generation={1}
      onClick={() => console.log("OK")}
      colorEmpty="#ececec"
      colorGamma="lightgreen"
      probe={probe}
    />
  );
  const div = screen.getByTestId("R1:C1");
  fireEvent.mouseOver(div, { buttons: 1 });
  expect(probe).toBeCalled();
});
it("Cell mouse over event without pressed button", () => {
  const probe = jest.fn();
  render(
    <Cell
      row={1}
      col={1}
      generation={1}
      onClick={() => console.log("OK")}
      colorEmpty="#ececec"
      colorGamma="lightgreen"
      probe={probe}
    />
  );
  const div = screen.getByTestId("R1:C1");
  fireEvent.mouseOver(div);
  expect(probe).not.toBeCalled();
});
