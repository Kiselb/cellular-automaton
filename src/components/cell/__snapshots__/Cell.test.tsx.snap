// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Cell renders correctly 1`] = `
<DocumentFragment>
  <div
    class="cell"
    data-testid="R1:C1"
    style="background: rgb(255, 255, 0);"
  />
</DocumentFragment>
`;
